File: src/main/java/gov/nih/nlm/ArangoDbUtilities.java
package gov.nih.nlm;

import com.arangodb.*;
import com.arangodb.entity.EdgeDefinition;
import com.arangodb.entity.arangosearch.AnalyzerFeature;
import com.arangodb.entity.arangosearch.CollectionLink;
import com.arangodb.entity.arangosearch.FieldLink;
import com.arangodb.entity.arangosearch.StoreValuesType;
import com.arangodb.entity.arangosearch.analyzer.*;
import com.arangodb.model.EdgeCollectionRemoveOptions;
import com.arangodb.model.VertexCollectionRemoveOptions;
import com.arangodb.model.arangosearch.ArangoSearchCreateOptions;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import com.arangodb.entity.CollectionEntity;
import com.arangodb.entity.CollectionType;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

/**
 * Provides utilities for managing named ArangoDB databases, graphs, vertex
 * collections, and edge collections.
 *
 * @author Raymond LeClair
 */
public class ArangoDbUtilities {

    /**
     * An ArangoDB instance
     */
    public final ArangoDB arangoDB;

    /**
     * Build the ArangoDB instance specified in the system environment.
     */
    public ArangoDbUtilities() {
        Map<String, String> env = System.getenv();
        arangoDB = new ArangoDB.Builder().host(env.get("ARANGO_DB_HOST"), Integer.parseInt(env.get("ARANGO_DB_PORT"))).user(env.get("ARANGO_DB_USER")).password(env.get("ARANGO_DB_PASSWORD")).build();
    }

    /**
     * Build the ArangoDB instance specified in the provided environment.
     *
     * @param env Environment map
     */
    public ArangoDbUtilities(Map<String, String> env) {
        arangoDB = new ArangoDB.Builder().host(env.get("ARANGO_DB_HOST"), Integer.parseInt(env.get("ARANGO_DB_PORT"))).user(env.get("ARANGO_DB_USER")).password(env.get("ARANGO_DB_PASSWORD")).build();
    }

    /**
     * Exercise the utilities.
     *
     * @param args (None expected)
     */
    public static void main(String[] args) {
        ArangoDbUtilities arangoDbUtilities = new ArangoDbUtilities();
        String databaseName = "myDb";
        ArangoDatabase db = arangoDbUtilities.createOrGetDatabase(databaseName);
        String graphName = "myGraph";
        ArangoGraph graph = arangoDbUtilities.createOrGetGraph(db, graphName);
        String vertexOneName = "myVertexOne";
        arangoDbUtilities.createOrGetVertexCollection(graph, vertexOneName);
        String vertexTwoName = "myVertexTwo";
        arangoDbUtilities.createOrGetVertexCollection(graph, vertexTwoName);
        ArangoEdgeCollection edgeCollection = arangoDbUtilities.createOrGetEdgeCollection(graph, vertexOneName, vertexTwoName);
        arangoDbUtilities.deleteEdgeCollection(graph, edgeCollection.name());
        arangoDbUtilities.deleteVertexCollection(graph, vertexTwoName);
        arangoDbUtilities.deleteVertexCollection(graph, vertexOneName);
        arangoDbUtilities.deleteGraph(db, graphName);
        arangoDbUtilities.deleteDatabase(databaseName);
    }

    /**
     * Create or get a named database.
     *
     * @param databaseName Name of the database to create or get
     * @return Named database
     */
    public ArangoDatabase createOrGetDatabase(String databaseName) {
        // Create the database, if needed
        if (!arangoDB.db(databaseName).exists()) {
            System.out.println("Creating database: " + databaseName);
            if (!arangoDB.createDatabase(databaseName)) {
                throw new RuntimeException("Could not create database: " + databaseName);
            }
        }
        // Get the database
        System.out.println("Getting database: " + databaseName);
        return arangoDB.db(databaseName);
    }

    /**
     * Delete a named database.
     *
     * @param databaseName Name of the database to delete
     */
    public void deleteDatabase(String databaseName) {
        // Delete the database, if needed
        if (arangoDB.db(databaseName).exists()) {
            System.out.println("Deleting database: " + databaseName);
            if (!arangoDB.db(databaseName).drop()) {
                throw new RuntimeException("Could not delete database: " + databaseName);
            }
        }
    }

    /**
     * Create or get a named graph.
     *
     * @param db        Database in which to create or get the graph
     * @param graphName Name of the graph to create or get
     * @return Named graph
     */
    public ArangoGraph createOrGetGraph(ArangoDatabase db, String graphName) {
        // Create the graph, if needed
        if (!db.graph(graphName).exists()) {
            System.out.println("Creating graph: " + graphName);
            Collection<EdgeDefinition> edgeDefinitions = new ArrayList<>();
            db.createGraph(graphName, edgeDefinitions);
        }
        // Get the graph
        System.out.println("Getting graph: " + graphName);
        return db.graph(graphName);
    }

    /**
     * Delete a named graph.
     *
     * @param db        Database in which to delete the graph
     * @param graphName Name of the graph to delete
     */
    public void deleteGraph(ArangoDatabase db, String graphName) {
        // Delete the graph, if needed
        if (db.graph(graphName).exists()) {
            System.out.println("Deleting graph: " + graphName);
            db.graph(graphName).drop();
        }
    }

    /**
     * Create or get a named vertex collection.
     *
     * @param graph      Graph in which to create or get the vertex collection
     * @param vertexName Name of the vertex collection to create or get
     * @return Named vertex collection
     */
    public ArangoVertexCollection createOrGetVertexCollection(ArangoGraph graph, String vertexName) {
        // Create the vertex collection, if needed
        if (!graph.getVertexCollections().contains(vertexName)) {
            System.out.println("Creating vertex collection: " + vertexName);
            graph.addVertexCollection(vertexName);
        }
        // Get the vertex collection
        System.out.println("Getting vertex collection: " + vertexName);
        return graph.vertexCollection(vertexName);
    }

    /**
     * Delete a named vertex collection.
     *
     * @param graph      Graph in which to delete the vertex collection
     * @param vertexName Name of the vertex collection to delete
     */
    public void deleteVertexCollection(ArangoGraph graph, String vertexName) {
        // Delete the vertex collection, if needed
        if (graph.getVertexCollections().contains(vertexName)) {
            System.out.println("Deleting vertex collection: " + vertexName);
            VertexCollectionRemoveOptions options = new VertexCollectionRemoveOptions();
            options.dropCollection(true);
            graph.vertexCollection(vertexName).remove(options);
        }
    }

    /**
     * Create, or get a named edge collection from and to the named vertices.
     *
     * @param graph          Graph in which to create, or get the edge collection
     * @param fromVertexName Name of the vertex collection from which the edge
     *                       originates
     * @param toVertexName   Name of the vertex collection to which the edge
     *                       terminates
     * @return Named edge collection
     */
    public ArangoEdgeCollection createOrGetEdgeCollection(ArangoGraph graph, String fromVertexName, String toVertexName) {
        // Create edge collection, if needed
        String collectionName = fromVertexName + "-" + toVertexName;
        if (!graph.getEdgeDefinitions().contains(collectionName)) {
            System.out.println("Creating edge collection: " + collectionName);
            EdgeDefinition edgeDefinition = new EdgeDefinition().collection(collectionName).from(fromVertexName).to(toVertexName);
            graph.addEdgeDefinition(edgeDefinition);
        }
        // Get the edge collection
        System.out.println("Getting edge collection: " + collectionName);
        return graph.edgeCollection(collectionName);
    }

    /**
     * Delete a named edge collection.
     *
     * @param graph    Graph in which to create, or get the edge collection
     * @param edgeName Name of the edge collection to delete
     */
    public void deleteEdgeCollection(ArangoGraph graph, String edgeName) {
        // Delete the edge collection, if needed
        if (graph.getEdgeDefinitions().contains(edgeName)) {
            System.out.println("Deleting edge collection: " + edgeName);
            EdgeCollectionRemoveOptions options = new EdgeCollectionRemoveOptions();
            options.dropCollections(true);
            graph.edgeCollection(edgeName).remove(options);
        }
    }

    /**
     * Print a summary of document counts for each vertex and edge
     * collection in the specified database.
     *
     * @param db Database to summarize
     * @return Map with "vertex" and "edge" keys, each mapping collection
     *         names to document counts (sorted alphabetically)
     */
    public Map<String, Map<String, Long>> printSummary(ArangoDatabase db) {
        Collection<CollectionEntity> collections = db.getCollections();
        List<CollectionEntity> vertexCollections = new ArrayList<>();
        List<CollectionEntity> edgeCollections = new ArrayList<>();
        for (CollectionEntity collection : collections) {
            if (collection.getIsSystem()) {
                continue;
            }
            if (collection.getType() == CollectionType.EDGES) {
                edgeCollections.add(collection);
            } else {
                vertexCollections.add(collection);
            }
        }
        vertexCollections.sort(Comparator.comparing(CollectionEntity::getName));
        edgeCollections.sort(Comparator.comparing(CollectionEntity::getName));

        Map<String, Long> vertexCounts = new LinkedHashMap<>();
        System.out.println("Vertex collections:");
        long vertexTotal = 0;
        for (CollectionEntity collection : vertexCollections) {
            long count = db.collection(collection.getName()).count().getCount();
            vertexTotal += count;
            vertexCounts.put(collection.getName(), count);
            System.out.printf("  %-40s %,d%n", collection.getName(), count);
        }
        System.out.printf("  %-40s %,d%n", "TOTAL", vertexTotal);

        Map<String, Long> edgeCounts = new LinkedHashMap<>();
        System.out.println("Edge collections:");
        long edgeTotal = 0;
        for (CollectionEntity collection : edgeCollections) {
            long count = db.collection(collection.getName()).count().getCount();
            edgeTotal += count;
            edgeCounts.put(collection.getName(), count);
            System.out.printf("  %-40s %,d%n", collection.getName(), count);
        }
        System.out.printf("  %-40s %,d%n", "TOTAL", edgeTotal);

        Map<String, Map<String, Long>> summary = new LinkedHashMap<>();
        summary.put("vertex", vertexCounts);
        summary.put("edge", edgeCounts);
        return summary;
    }

    /**
     * Create n-gram and text analyzers in the specified database.
     *
     * @param db Database in which to create the analyzers
     */
    public void createAnalyzers(ArangoDatabase db) {
        // Create n-gram analyzer
        NGramAnalyzer ngramAnalyzer = new NGramAnalyzer();
        ngramAnalyzer.setName("n-gram");
        ngramAnalyzer.setFeatures(Set.of(AnalyzerFeature.frequency, AnalyzerFeature.position, AnalyzerFeature.norm));
        NGramAnalyzerProperties ngramProps = new NGramAnalyzerProperties();
        ngramProps.setMin(3);
        ngramProps.setMax(4);
        ngramProps.setPreserveOriginal(true);
        ngramProps.setStreamType(StreamType.utf8);
        ngramAnalyzer.setProperties(ngramProps);
        System.out.println("Creating analyzer: n-gram");
        db.createSearchAnalyzer(ngramAnalyzer);

        // Create text analyzer without stemming
        TextAnalyzer textAnalyzer = new TextAnalyzer();
        textAnalyzer.setName("text_en_no_stem");
        textAnalyzer.setFeatures(Set.of(AnalyzerFeature.frequency, AnalyzerFeature.position, AnalyzerFeature.norm));
        TextAnalyzerProperties textProps = new TextAnalyzerProperties();
        textProps.setLocale("en");
        textProps.setAnalyzerCase(SearchAnalyzerCase.lower);
        textProps.setAccent(false);
        textProps.setStemming(false);
        EdgeNgram edgeNgram = new EdgeNgram();
        edgeNgram.setMin(3);
        edgeNgram.setMax(12);
        edgeNgram.setPreserveOriginal(true);
        textProps.setEdgeNgram(edgeNgram);
        textAnalyzer.setProperties(textProps);
        System.out.println("Creating analyzer: text_en_no_stem");
        db.createSearchAnalyzer(textAnalyzer);
    }

    /**
     * Delete n-gram and text analyzers in the specified database.
     *
     * @param db Database in which to delete the analyzers
     */
    public void deleteAnalyzers(ArangoDatabase db) {
        try {
            System.out.println("Deleting analyzer: n-gram");
            db.deleteSearchAnalyzer("n-gram");
        } catch (ArangoDBException e) {
            System.out.println("Analyzer n-gram not found, skipping");
        }
        try {
            System.out.println("Deleting analyzer: text_en_no_stem");
            db.deleteSearchAnalyzer("text_en_no_stem");
        } catch (ArangoDBException e) {
            System.out.println("Analyzer text_en_no_stem not found, skipping");
        }
    }

    /**
     * Create an arangosearch view named "indexed" in the specified database,
     * using the collection maps JSON file to configure per-collection links.
     *
     * @param db                 Database in which to create the view
     * @param collectionMapsPath Path to the JSON file containing collection maps
     * @throws IOException if the JSON file cannot be read
     */
    public void createView(ArangoDatabase db, Path collectionMapsPath) throws IOException {
        // Read collection maps JSON
        ObjectMapper mapper = new ObjectMapper();
        JsonNode root = mapper.readTree(Files.readString(collectionMapsPath));
        JsonNode maps = root.get("maps");

        // Build view options with per-collection links
        ArangoSearchCreateOptions options = new ArangoSearchCreateOptions().commitIntervalMsec(1000L).consolidationIntervalMsec(1000L).cleanupIntervalStep(2L);

        for (JsonNode collectionMap : maps) {
            String vertexName = collectionMap.get(0).asText();
            // Skip non-vertex entries
            if (vertexName.equals("edges") || vertexName.equals("TEST_DOCUMENT_COLLECTION") || vertexName.equals("TEST_EDGE_COLLECTION")) {
                continue;
            }

            // Collect field names
            JsonNode individualFields = collectionMap.get(1).get("individual_fields");
            List<FieldLink> fieldLinks = new ArrayList<>();
            for (JsonNode field : individualFields) {
                String fieldName = field.get("field_to_display").asText();
                fieldLinks.add(FieldLink.on(fieldName).analyzers("text_en", "text_en_no_stem", "n-gram", "identity"));
            }

            // Create collection link
            CollectionLink link = CollectionLink.on(vertexName).analyzers("identity").includeAllFields(false).storeValues(StoreValuesType.NONE).trackListPositions(false).fields(fieldLinks.toArray(new FieldLink[0]));

            options.link(link);
        }

        System.out.println("Creating view: indexed");
        db.createArangoSearch("indexed", options);
    }

    /**
     * Delete the arangosearch view named "indexed" in the specified database.
     *
     * @param db Database in which to delete the view
     */
    public void deleteView(ArangoDatabase db) {
        try {
            System.out.println("Deleting view: indexed");
            db.arangoSearch("indexed").drop();
        } catch (ArangoDBException e) {
            System.out.println("View indexed not found, skipping");
        }
    }
}

File: python/src/ArangoDbUtilities.py
import json
import os
from pprint import pprint
import random

from arango import ArangoClient

ARANGO_URL = "http://localhost:8529"
ARANGO_CLIENT = ArangoClient(hosts=ARANGO_URL)
ARANGO_ROOT_PASSWORD = os.getenv("ARANGO_DB_PASSWORD", "")
SYS_DB = ARANGO_CLIENT.db("_system", username="root", password=ARANGO_ROOT_PASSWORD)


def create_or_get_database(database_name):
    """Create or get an ArangoDB database.

    Parameters
    ----------
    database_name : str
        Name of the database to create or get

    Returns
    -------
    db : arango.database.StandardDatabase
        Database
    """
    # Create database, if needed
    if not SYS_DB.has_database(database_name):
        print(f"Creating ArangoDB database: {database_name}")
        SYS_DB.create_database(database_name)

    # Connect to database
    print(f"Getting ArangoDB database: {database_name}")
    db = ARANGO_CLIENT.db(database_name, username="root", password=ARANGO_ROOT_PASSWORD)

    return db


def delete_database(database_name):
    """Delete an ArangoDB database.

    Parameters
    ----------
    database_name : str
        Name of the database to delete

    Returns
    -------
    None
    """
    # Delete database, if needed
    if SYS_DB.has_database(database_name):
        print(f"Deleting ArangoDB database: {database_name}")
        SYS_DB.delete_database(database_name)


def create_or_get_graph(db, graph_name):
    """Create or get an ArangoDB database graph.

    Parameters
    ----------
    db : arango.database.StandardDatabase
        Database
    graph_name : str
        Name of the graph to create or get

    Returns
    -------
    graph : arango.graph.Graph
        Database graph
    """
    # Create, or get the graph
    if not db.has_graph(graph_name):
        print(f"Creating database graph: {graph_name}")
        graph = db.create_graph(graph_name)
    else:
        print(f"Getting database graph: {graph_name}")
        graph = db.graph(graph_name)

    return graph


def delete_graph(db, graph_name):
    """Delete an ArangoDB database graph.

    Parameters
    ----------
    db : arango.database.StandardDatabase
        Database
    graph_name : str
        Name of the graph to delete

    Returns
    -------
    None
    """
    # Delete the graph
    if db.has_graph(graph_name):
        print(f"Deleting database graph: {graph_name}")
        db.delete_graph(graph_name)


def create_or_get_vertex_collection(graph, vertex_name):
    """Create, or get an ArangoDB database graph vertex collection.

    Parameters
    ----------
    graph : arango.graph.Graph
        Graph
    vertex_name : str
        Name of the vertex collection to create or get

    Returns
    -------
    collection : arango.collection.VertexCollection
        Graph vertex collection
    """
    # Create, or get the vertex collection
    if not graph.has_vertex_collection(vertex_name):
        print(f"Creating graph vertex collection: {vertex_name}")
        collection = graph.create_vertex_collection(vertex_name)
    else:
        print(f"Getting graph vertex collection: {vertex_name}")
        collection = graph.vertex_collection(vertex_name)

    return collection


def delete_vertex_collection(graph, vertex_name):
    """Delete an ArangoDB database graph vertex collection.

    Parameters
    ----------
    graph : arango.graph.Graph
        Graph
    vertex_name : str
        Name of the vertex collection to delete

    Returns
    -------
    None
    """
    # Delete the vertex collection
    if graph.has_vertex_collection(vertex_name):
        print(f"Deleting graph vertex collection: {vertex_name}")
        graph.delete_vertex_collection(vertex_name)


def create_or_get_edge_collection(graph, from_vertex_name, to_vertex_name):
    """Create, or get an ArangoDB database edge collection from and
    to the specified vertices.

    Parameters
    ----------
    graph : arango.graph.Graph
        Graph
    from_vertex : str
        Name of the vertex collection from which the edge originates
    to_vertex : str
        Name of the vertex collection to which the edge terminates

    Returns
    -------
    collection : arango.collection.EdgeCollection
        Graph edge collection
    collection_name : str
        Name of the edge collection
    """
    # Create, or get the edge collection
    collection_name = f"{from_vertex_name}-{to_vertex_name}"
    if not graph.has_edge_definition(collection_name):
        print(f"Creating edge definition: {collection_name}")
        collection = graph.create_edge_definition(
            edge_collection=collection_name,
            from_vertex_collections=[f"{from_vertex_name}"],
            to_vertex_collections=[f"{to_vertex_name}"],
        )
    else:
        print(f"Getting edge collection: {collection_name}")
        collection = graph.edge_collection(collection_name)

    return collection, collection_name


def delete_edge_collection(graph, edge_name):
    """Delete an ArangoDB database graph edge definition and collection.

    Parameters
    ----------
    graph : arango.graph.Graph
        Graph
    edge_name : str
        Name of the edge definition and collection to delete

    Returns
    -------
    None
    """
    # Delete the collection
    if graph.has_edge_definition(edge_name):
        print(f"Deleting graph edge definition and collection: {edge_name}")
        graph.delete_edge_definition(edge_name)


def print_summary(database_name):
    """Print a summary of document counts for each vertex and edge collection
    in the named database.

    Parameters
    ----------
    database_name : str
        Name of the database to summarize

    Returns
    -------
    summary : dict
        Dictionary with "vertex" and "edge" keys, each mapping collection
        names to document counts (sorted alphabetically)
    """
    db = create_or_get_database(database_name)
    vertex_collections = []
    edge_collections = []
    for collection in db.collections():
        if collection["system"]:
            continue
        if collection["type"] == "edge":
            edge_collections.append(collection["name"])
        else:
            vertex_collections.append(collection["name"])

    vertex_counts = {}
    print("Vertex collections:")
    vertex_total = 0
    for name in sorted(vertex_collections):
        count = db.collection(name).count()
        vertex_total += count
        vertex_counts[name] = count
        print(f"  {name:<40s} {count:,d}")
    print(f"  {'TOTAL':<40s} {vertex_total:,d}")

    edge_counts = {}
    print("Edge collections:")
    edge_total = 0
    for name in sorted(edge_collections):
        count = db.collection(name).count()
        edge_total += count
        edge_counts[name] = count
        print(f"  {name:<40s} {count:,d}")
    print(f"  {'TOTAL':<40s} {edge_total:,d}")

    return {"vertex": vertex_counts, "edge": edge_counts}


def create_analyzers(database_name):
    """Create n-gram and text analyzers in the named database.

    Parameters
    ----------
    database_name : str
        Name of the database in which to create the analyzers

    Returns
    -------
    None
    """
    db = create_or_get_database(database_name)
    db.create_analyzer(
        name=f"n-gram",
        analyzer_type="ngram",
        properties={
            "min": 3,
            "max": 4,
            "preserveOriginal": True,
            "streamType": "utf8",
            "startMarker": "",
            "endMarker": "",
        },
        features=["frequency", "position", "norm"],
    )
    db.create_analyzer(
        name=f"text_en_no_stem",
        analyzer_type="text",
        properties={
            "locale": "en",
            "case": "lower",
            "accent": False,
            "stemming": False,
            "edgeNgram": {
                "min": 3,
                "max": 12,
                "preserveOriginal": True,
            },
        },
        features=["frequency", "position", "norm"],
    )


def delete_analyzers(database_name):
    """Delete n-gram and text analyzers in the named database.

    Parameters
    ----------
    database_name : str
        Name of the database in which to delete the analyzer

    Returns
    -------
    None
    """
    db = create_or_get_database(database_name)
    db.delete_analyzer(f"{database_name}::n-gram", ignore_missing=True)
    db.delete_analyzer(f"{database_name}::text_en_no_stem", ignore_missing=True)


def create_view(database_name, collection_maps_name):
    """Create views in the Evidence and Knowledge graphs in the
    specified database based on the specified collection_maps JSON
    file.

    Parameters
    ----------
    database_name : str
        Name of the database in which to create the views
    collection_maps_name : str
        Name of the JSON file containing a mapping for each vertex
        collection

    Returns
    -------
    None
    """
    # Populate the view properties from the collections map
    with open(collection_maps_name, "r") as fp:
        collection_maps = json.load(fp)
    properties = {
        "writebufferSizeMax": 33554432,
        "id": "74447522",
        "storedValues": [],
        "consolidationPolicy": {
            "type": "tier",
            "segmentsBytesFloor": 2097152,
            "segmentsBytesMax": 5368709120,
            "segmentsMax": 10,
            "segmentsMin": 1,
            "minScore": 0,
        },
        "writebufferActive": 0,
        "links": {},
        "commitIntervalMsec": 1000,
        "consolidationIntervalMsec": 1000,
        "globallyUniqueId": "h1D09A664A5DB/74447522",
        "cleanupIntervalStep": 2,
        "primarySort": [],
        "primarySortCompression": "lz4",
        "writebufferIdle": 64,
    }
    for collection_map in collection_maps["maps"]:
        vertex_name = collection_map[0]
        # TODO: Restructure collection_maps to separate vertices and edges
        if vertex_name in ["edges", "TEST_DOCUMENT_COLLECTION", "TEST_EDGE_COLLECTION"]:
            continue
        vertex_fields = [
            d["field_to_display"] for d in collection_map[1]["individual_fields"]
        ]
        properties["links"][vertex_name] = {}
        properties["links"][vertex_name]["analyzers"] = ["identity"]
        properties["links"][vertex_name]["fields"] = {}
        for vertex_field in vertex_fields:
            properties["links"][vertex_name]["fields"][vertex_field] = {
                "analyzers": ["text_en", "text_en_no_stem", "n-gram", "identity"]
            }
        properties["links"][vertex_name]["includeAllFields"] = False
        properties["links"][vertex_name]["storeValues"] = "none"
        properties["links"][vertex_name]["trackListPositions"] = False

    db = create_or_get_database(database_name)
    if database_name == "Cell-KN-Phenotypes":
        keys = list(properties["links"].keys())
        for key in keys:
            if key not in [
                "BGS",
                "BMC",
                "CHEMBL",
                "CL",
                "CS",
                "CSD",
                "GO",
                "GS",
                "MONDO",
                "NCBITaxon",
                "PATO",
                "PUB",
                "PR",
                "RS",
                "UBERON",
            ]:
                del properties["links"][key]

    db.create_view(
        name="indexed",
        view_type="arangosearch",
        properties=properties,
    )


def delete_view(database_name):
    """Delete views in the Evidence and Knowledge graphs in the
    specified database.

    Parameters
    ----------
    database_name : str
        Name of the database in which to create the views

    Returns
    -------
    None
    """
    db = create_or_get_database(database_name)
    db.delete_view("indexed")


def print_vertex_examples(database_name, graph_name):
    """Delete the bigram analyzer in the named database.

    Parameters
    ----------
    database_name : str
        Name of the database in which to delete the analyzer
    graph_name : str
        Name of the graph to create or get

    Returns
    -------
    None
    """
    # Get the database and graph
    db = create_or_get_database(database_name)
    graph = create_or_get_graph(db, graph_name)

    # Collect relevant vertex names
    vertex_names = []
    for collection in db.collections():
        if collection["type"] != "document" or collection["name"][0] == "_":
            continue
        vertex_names.append(collection["name"])

    # Select one vertex randomly
    random.seed(a=0, version=2)
    for vertex_name in sorted(vertex_names):
        vertex_collection = create_or_get_vertex_collection(graph, vertex_name)
        vertex_keys = list(vertex_collection.keys())
        vertex_key = vertex_keys[random.randint(0, vertex_collection.count() - 1)]
        vertex = vertex_collection.get(vertex_key)
        print()
        print(vertex_name)
        print()
        pprint(vertex)
